### DISCUSSION TITLE
Cannot cache when using Debouncer

### DISCUSSION THREAD

**Discussion by iamsahilsonawane** [rrousselGit/riverpod#1687] [Q&A] [answered]
I am trying to implement pagination with Riverpod as per the pub example. I wanted to implement debouncing with the delayed method in the example and every other solution I could find. 

So the code looks something like this:
```dart
final paginatedVendorListProvider =
    FutureProvider.family<PaginatedResponse<Vendor>, SearchPaginationArgs>(
        (ref, pageData) async {
  ref.watch(refreshVendorListProviders);
  final cancelToken = ref.cancelToken();

  final adminRepo = ref.watch(adminRepositoryProvider);

  await ref.debounce(const Duration(seconds: 1));
  if (cancelToken.isCancelled) {
    throw Exception('cancelled');
  }

  return await adminRepo.fetchAllVendors(
      pageNumber: pageData.page!,
      search: pageData.search!,
      cancelToken: cancelToken);
});
```
Still, the `adminRepo.fetchAllVendors`, if I type in their 3 letters, it waits for 1 second and then calls the method 3 times after. This behavior doesn't happen in the pub example though. I am doing something wrong?

**Comment by TimWhiting** [comment, marked as answer]
The issue is that you are using a family with the arguments. A family creates a different provider per search, and since you are not creating them as autoDispose providers they are never being disposed and not canceling. 

There are two solutions, you can create a StateProvider to store the arguments, and then watch that provider in this one (this gets rid of the family aspect of this provider). Or you can make the family autoDispose as well. The code generation make providers autoDispose by default, which is why the examples behavior is different.

**Reply by iamsahilsonawane** [reply]
I see.. I'll first try both out and be back on this. 

Thank you

**Reply by iamsahilsonawane** [reply]
@TimWhiting When I make the provider `autoDispose`, the pagination works absolutely fine but the pages are fetched over and over each time user scrolls up or down, basically it's not caching for obvious reasons. (This is also reproducible in the example apps)

Also, I don't quite understand how would I do the first solution (i.e. to use `StateProvider` and watch that which removes the family), as each time the provider gets created each batch gets its own independent args. How can I generalize that to use just one arg?

What I want is simple pagination wherein results get cached when they're fetched and for search, it should debounce searches.

**Reply by rrousselGit** [reply]
You're looking for #1664

**Reply by iamsahilsonawane** [reply]
@rrousselGit It cannot find the `keepAlive()` method on `Ref`.  Nor does it work on the `pub` example.

I am using `v2.0.0-dev.9`

Thanks

**Reply by iamsahilsonawane** [reply]
I think the example is incorrect and will work on the disposable providers. So that should be:

```dart
extension on AutoDisposeRef {
  // When invoked keeps your provider alive for [duration]
  void cacheFor(Duration duration) {
    final link = keepAlive();
    final timer = Timer(duration, () => link.close());
    onDispose(() => timer.cancel);
  }
}

final myProvider = Provider.autoDispose<int>((ref) {
  ref.cacheFor(Duration(minutes: 5));
});
```

**Reply by rrousselGit** [reply]
You're right.

**Reply by iamsahilsonawane** [reply]
Now with this, again the scroll works fine, but now the search has the debouncing search issue. That's because the provider is not getting disposed and therefore the debouncing doesn't work. Leading to the same problem as previous

**Reply by rrousselGit** [reply]
It is getting disposed. In 5 minutes that is.

**Reply by rrousselGit** [reply]
If you want to abort the network-request before the duration completes, you can use `ref.onCancel`/`ref.onResume`
These are life-cycles for when a provider stops being used.

Like:

```dart
var debouncing = true;
ref.onCancel(() {
  if (debouncing) cancelToken.cancel();
});
await Future.delayed(duration);
debouncing = false;

if (cancelToken.cancelled) {
  // abort
}
```

But that's why https://github.com/rrousselGit/riverpod/issues/1664 is open. That issue is about implementing that for you

**Reply by iamsahilsonawane** [reply]
With this implementation, it works fine sometimes, but then I start getting these:
```shell
Unhandled Exception: Bad state: The provider AutoDisposeFutureProvider<PaginatedResponse<Vendor>>#0b102(SearchPaginationArgs(1, ma)) was disposed before a value was emitted.
```

Implementation: 

```dart
final paginatedVendorListProvider = FutureProvider.autoDispose
    .family<PaginatedResponse<Vendor>, SearchPaginationArgs>(
        (ref, pageData) async {
  ref.cacheFor(const Duration(minutes: 5));

  ref.watch(refreshVendorListProviders);

  final adminRepo = ref.watch(adminRepositoryProvider);

  final cancelToken = ref.cancelToken();
  var debouncing = true;
  ref.onCancel(() {
    if (debouncing) cancelToken.cancel();
  });
  await ref.debounce(const Duration(seconds: 1));
  debouncing = false;

  if (cancelToken.isCancelled) {
    throw Exception('cancelled');
  }

  log("Fetching from API");

  return await adminRepo.fetchAllVendors(
      pageNumber: pageData.page!,
      search: pageData.search!,
      cancelToken: cancelToken);
});
```

**Reply by iamsahilsonawane** [reply]
From what I understand, there are two issues that are related to each other.

1st is that for pagination scroll with cache to work it needs to be either a non-disposable provider or we can use the `keepAlive` to cache. With these solutions, it works fine.

2nd is that for search to have a debouncer, it has to be disposable. Or the second solution you proposed (with the `onCancel`) which honestly I didn't quite understand very well.

So basically I am stuck between these two cases to be in a single solution. Achieving one is fine and works really well.

Thanks for the continued support!

**Reply by iamsahilsonawane** [reply]
**Update:** Sorry I didn't read the full doc for `onCancel` well and now I've removed the cache time and converted the provider to be a non-disposable family provider again. And now the network request cancellation works!

Thank you @rrousselGit and @TimWhiting for your support on this matter.

**Reply by iamsahilsonawane** [reply]
Though I have another question on the pagination and search now that I can see 🥲 

But that's a little different topic so, I'll create a new question for that?

**Reply by TimWhiting** [reply]
Yes create a new question, so others can find the questions and answers easier
